src/controllers/krate/update.rs

```diff
@@ -0,0 +1,199 @@
+    app: AppState,
+    path: CratePath,
+    req: Parts,
+    Extension(real_ip): Extension<RealIp>,
+    Json(body): Json<PatchRequest>,
+) -> AppResult<Json<PatchResponse>> {
+    let mut conn = app.db_write().await?;
+
+    // Check that the user is authenticated
+    let auth = AuthCheck::only_cookie().check(&req, &mut conn).await?;
+
+    // Check that the crate exists
+    let krate = path.load_crate(&mut conn).await?;
+
+    // Update crate settings in a transaction
+    conn.transaction(|conn| {
+        update_inner(conn, &app, &krate, auth.user(), &real_ip, body).scope_boxed()
+    })
+    .await
+}
+
+async fn update_inner(
+    conn: &mut diesel_async::AsyncPgConnection,
+    app: &AppState,
+    krate: &Crate,
+    user: &User,
+    real_ip: &RealIp,
+    body: PatchRequest,
+) -> AppResult<Json<PatchResponse>> {
+    // Query user owners to check permissions and send emails
+    let user_owners = crate_owners::table
+        .inner_join(users::table)
+        .inner_join(emails::table.on(users::id.eq(emails::user_id)))
+        .filter(crate_owners::crate_id.eq(krate.id))
+        .filter(crate_owners::deleted.eq(false))
+        .filter(crate_owners::owner_kind.eq(crate::models::OwnerKind::User))
+        .select((users::id, users::gh_login, emails::email, emails::verified))
+        .load::<(i32, String, String, bool)>(conn)
+        .await?;
+
+    // Check that the authenticated user is an owner
+    if !user_owners.iter().any(|(id, _, _, _)| *id == user.id) {
+        let msg = "only owners have permission to modify crate settings";
+        return Err(custom(StatusCode::FORBIDDEN, msg));
+    }
+
+    // Update trustpub_only if provided
+    if let Some(trustpub_only) = body.trustpub_only {
+        diesel::update(crates::table)
+            .filter(crates::id.eq(krate.id))
+            .set(crates::trustpub_only.eq(trustpub_only))
+            .execute(conn)
+            .await?;
+
+        // Audit log the setting change
+        info!(
+            target: "audit",
+            action = "trustpub_only_change",
+            krate.name = %krate.name,
+            network.client.ip = %**real_ip,
+            usr.id = user.id,
+            usr.name = %user.gh_login,
+            "User {} set trustpub_only={trustpub_only} for crate {}",
+            user.gh_login,
+            krate.name
+        );
+
+        // Send email notifications to all crate owners
+        for (_, gh_login, email_address, email_verified) in &user_owners {
+            if *email_verified {
+                let email = TrustpubOnlyChangedEmail {
+                    recipient: gh_login,
+                    auth_user: user,
+                    krate,
+                    trustpub_only,
+                };
+
+                if let Err(err) = email.send(app, email_address).await {
+                    warn!("Failed to send trustpub_only notification to {email_address}: {err}");
+                }
+            }
+        }
+    }
```

Comment:
Not a blocker, but I'm wondering if we should check whether the given value is different from the existing value since we are sending a notification?

---

Nevertheless, I especially love the introduced notification mechanism here. Without it, a malicious user with a leaked API token could change the setting and publish a malicious version, which the crate owner might not notice. But with this mechanism, the crate owner has a chance to rotate the token and make amends!